src/Build/BackEnd/Shared/BuildRequestConfiguration.cs

```diff
@@ -710,6 +714,47 @@ public List<string> GetAfterTargetsForDefaultTargets(BuildRequest request)
             return null;
         }
 
+        private Func<string, bool> shouldSkipStaticGraphIsolationOnReference;
+
+        public bool ShouldSkipIsolationConstraintsForReference(string referenceFullPath)
+        {
+            ErrorUtilities.VerifyThrowInternalNull(Project, nameof(Project));
+            ErrorUtilities.VerifyThrowInternalLength(referenceFullPath, nameof(referenceFullPath));
+            ErrorUtilities.VerifyThrow(Path.IsPathRooted(referenceFullPath), "Method does not treat path normalization cases");
+
+            if (shouldSkipStaticGraphIsolationOnReference == null)
+            {
+                shouldSkipStaticGraphIsolationOnReference = GetReferenceFilter();
+            }
+
+            return shouldSkipStaticGraphIsolationOnReference(referenceFullPath);
+
+            Func<string, bool> GetReferenceFilter()
+            {
+                lock (_syncLock)
+                {
+                    if (shouldSkipStaticGraphIsolationOnReference != null)
+                    {
+                        return shouldSkipStaticGraphIsolationOnReference;
+                    }
+
+                    var items = Project.GetItems(ItemTypeNames.ReferencesToSkipGraphIsolationConstraintsOn);
+
+                    if (items.Count == 0 || items.All(i => string.IsNullOrWhiteSpace(i.EvaluatedInclude)))
+                    {
+                        return _ => false;
+                    }
+
+                    var fragments = items.SelectMany(i => ExpressionShredder.SplitSemiColonSeparatedList(i.EvaluatedInclude));
+                    var glob = new CompositeGlob(
+                        fragments
+                            .Select(s => MSBuildGlob.Parse(Project.Directory, s)));
```

Comment:
@danmosemsft 
Copied from https://github.com/microsoft/msbuild/pull/4407#discussion_r288699254
> Is this perf sensitive code? Historically we kept LINQ out of the codebase, or at least out of perf sensitive places. It tends to have hidden costs.

Reply:
I think the extra boost to readability is worth it over the perf impact. The glob parsing is way more expensive than the link operation, and the code executes only once per BuildRequestConfiguration.